mewarpx/mewarpx/diags_store/particle_histogram_diagnostic.py

```diff
@@ -0,0 +1,181 @@
+        """Initialize diagnostic.
+
+        Arguments:
+            diag_steps (int): Number of steps between each output.
+            species_list (list of species): Species for which histograms should
+                be recorded. If None, all simulation species will be included.
+            assembly (mewarpx.Assembly object): The assembly object in which
+                particles are scraped.
+            linres (int or list of 3 ints): Number of bins to create along each
+                dimension.
+            plot (bool): If True plot the histograms at each diagnostic step.
+            kwargs: See :class:`mewarpx.diags_store.diag_base.WarpXDiagnostic`
+                for more timing options.
+        """
+        self.write_dir = os.path.join(self.DIAG_DIR, self.PHIST_DIAG_DIR)
+
+        self.species_list = species_list
+        if self.species_list is None:
+            self.species_list = mwxrun.simulation.species
+        self.assembly = assembly
+        if np.size(linres) != 3:
+            linres = [linres, linres, linres]
+        self.linres_x, self.linres_y, self.linres_z = linres
+        self.plot = plot
+
+        # set sensible linres based on simulation geometry - can override
+        # user given arguments if they don't make sense
+        if mwxrun.dim == 0:
+            logger.warn(
+                "Particle histograms in 1d are trivial so this diagnostic "
+                "will not be installed."
+            )
+            return
+        elif mwxrun.dim == 2:
+            self.linres_y = 1
+
+        if isinstance(self.assembly, assemblies.ZPlane):
+            self.linres_z = 1
+
+        super(ParticleHistDiag, self).__init__(diag_steps=diag_steps, **kwargs)
+
+        # for now we hard code this diagnostic to only save weight but this
+        # can be extended in the future
+        self.quantities = ['w']
+
+        # install the scraper diag function
+        self.assembly.add_diag_fn(
+            self._process_scraped_particles, ['x', 'y', 'z'] + self.quantities
+        )
+
+        # create bins and data array after the simulation has been initialized
+        # since the bin specifications are written to file
+        callbacks.installafterinit(self.initialize)
+
+        # install callback function to save and reset histogram
+        callbacks.installafterstep(self.save_and_reset_histogram)
+
+    def initialize(self):
+        """Set up the histogram bins and data array."""
+        self.setup_bins()
+        self.setup_array()
+
+    def setup_bins(self):
+        """Calculate the bin edges."""
+        domain = [(mwxrun.xmin, mwxrun.xmax), (mwxrun.ymin, mwxrun.ymax),
+                  (mwxrun.zmin, mwxrun.zmax)]
+        linres = [self.linres_x+1, self.linres_y+1, self.linres_z+1]
+
+        self.bins = []
+        for ii in range(3):
+            if linres[ii] <= 2:
+                self.bins.append([-np.inf, np.inf])
+            else:
+                self.bins.append(
```

Comment:
If you make this more generic, I would either optionally or always add the -inf to min and max to +inf bins

Reply:
I'll explain this in person.